YeelightAPI/DeviceLocator.cs

```diff
@@ -20,877 +20,881 @@
+
+        #endregion Depricated API. TODO: Remove
+
+        #region Async API Methods
+
+        /// <summary>
+        ///   Discover devices in LAN
+        /// </summary>
+        /// <returns>A <see cref="Task" /> that contains a collection of all discovered <see cref="Device" /> instances.</returns>
+        public static async Task<IEnumerable<Device>> DiscoverAsync() =>
+          await DeviceLocator.DiscoverAsync(deviceFoundReporter: null);
+
+        /// <summary>
+        ///   Discover devices in LAN
+        /// </summary>
+        /// <param name="cancellationToken">A <see cref="CancellationToken" /> to cancel the asynchronous operation.</param>
+        /// <returns>A <see cref="Task" /> that contains a collection of all discovered <see cref="Device" /> instances.</returns>
+        /// <exception cref="OperationCanceledException">Thrown when operation was cancelled by the caller.</exception>
+        public static async Task<IEnumerable<Device>> DiscoverAsync(CancellationToken cancellationToken) =>
+          await DeviceLocator.DiscoverAsync(deviceFoundReporter: null, cancellationToken);
+
+        /// <summary>
+        ///   Discover devices in LAN
+        /// </summary>
+        /// <param name="deviceFoundReporter">
+        ///   An implementation of <see cref="IProgress{T}" /> to handle discovered devices as soon
+        ///   they are available.
+        /// </param>
+        /// <returns>A <see cref="Task" /> that contains a collection of all discovered <see cref="Device" /> instances.</returns>
+        public static async Task<IEnumerable<Device>> DiscoverAsync(IProgress<Device> deviceFoundReporter) =>
+          await DeviceLocator.DiscoverAsync(deviceFoundReporter, CancellationToken.None);
+
+        /// <summary>
+        ///   Discover devices in LAN
+        /// </summary>
+        /// <param name="deviceFoundReporter">
+        ///   An implementation of <see cref="IProgress{T}" /> to handle discovered devices as soon
+        ///   they are available.
+        /// </param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken" /> to cancel the asynchronous operation.</param>
+        /// <returns>A <see cref="Task" /> that contains a collection of all discovered <see cref="Device" /> instances.</returns>
+        /// <exception cref="OperationCanceledException">Thrown when operation was cancelled by the caller.</exception>
+        public static async Task<IEnumerable<Device>> DiscoverAsync(
+          IProgress<Device> deviceFoundReporter,
+          CancellationToken cancellationToken)
+        {
+            IEnumerable<Task<DiscoveryResult>> tasks = NetworkInterface.GetAllNetworkInterfaces()
+              .Where(networkInterface => networkInterface.OperationalStatus == OperationalStatus.Up)
+              .Select(
+                networkInterface => DeviceLocator.SearchNetworkForDevicesAsync(networkInterface, deviceFoundReporter, cancellationToken));
```

Comment:
I think the call to `DeviceLocator.SearchNetworkForDevicesAsync` should be awaited for proper exception behavior. I missed this before. I have just realized this.

```
// Correct: async lambda expression
.Select(async networkInterface => await DeviceLocator.SearchNetworkForDevicesAsync(networkInterface, deviceFoundReporter, cancellationToken));
```